### PR TITLE
Add dynamic provisioner for cephfs

### LINKED ISSUE #1125: CephFS volumes handling via PVCs (dynamic provisioning)
Per #1115, it'd be interesting if CephFS volumes could be handled by storage classes and PVCs, I think this would make it a lot easier to deploy these volumes to helm charts for instance where people would just need to set the `storageClass` to something like `rook-filesystem` and then the `accessModes` to `ReadWriteMany` rather than hacking in YAML which is specific to Rook itself.

### LINKED ISSUE #3263: Dynamic provisioning for CephFS storage
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Currently, the ceph filesystem supports in Rook requires the user to create a external storage cephfs provisioner to provision cephfs volume that will be consumed by a pod. The user has to look up the IP address of the Mon IP address, run a external storagre provisioner pod , etc., which is a cumbersome experience. This could be improved by intergrating a dynamic provisioner in rook operator that would automiatically do those actions for the user and generate a PV without user intervention. All the user who wants to request a PVC needs to do is deploying ceph filesystem firstly and declare a PVC with a storageclass.

### PR DESCRIPTION
**Pull request by caisan**
This commit is inspired by:kubernetes-incubator/external-storage/ceph/cephfs.
Most code here is identical to that found in cephfs-provisioner.go of
external-storage/ceph/cephfs provisioner. Administrator can now use cephfs PVC
by just deploying filesystem and setting up cephfs storageclass.

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #1125 (#3263 has been closed as a duplicate of #1125)

**Checklist:**
- [x] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../CONTRIBUTING.md#comments)

### CODE REVIEW

**Comment by caisan** [comment]
@rohan47 @travisn @BlaineEXE   cloud you please review this PR ?

File: pkg/operator/ceph/operator.go [hunk 1]
```diff
@@ -143,6 +143,27 @@ func (o *Operator) Run() error {
 		logger.Infof("rook-provisioner %s started using %s flex vendor dir", name, vendor)
 	}
 
+	// Run dynamically cephFS provisioner
+	CephFSprovisionerName := "rook.io/filesystem"
```
Comment (rohan47) [inline pkg/operator/ceph/operator.go]: `ceph.rook.io/filesystem`

File: pkg/operator/ceph/provisioner/provisioner.go [hunk 2]
```diff
@@ -48,6 +68,16 @@ type RookVolumeProvisioner struct {
 	flexDriverVendor string
 }
 
+type cephFSProvisioner struct {
```
Comment (rohan47) [inline pkg/operator/ceph/provisioner/provisioner.go]: What about having a different code file for cephfs provisioner in the same package?

File: Documentation/ceph-filesystem.md [hunk 3]
```diff
@@ -139,6 +139,44 @@ After creating it with `kubectl create -f kube-registry.yaml`, you now have a do
...
 
+## Consume the Shared File System: K8s PVC
+
+In order to comsume shared file system by using PVC, you need create storageclass firstly. Here is the sample of storageclass:
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: rook-ceph-fs
+provisioner: rook.io/filesystem
```
[the remark this answers is not in the document]
Comment (caisan) [reply Documentation/ceph-filesystem.md]: agree.

File: pkg/operator/ceph/operator.go (hunk 1 above)
Comment (caisan) [reply pkg/operator/ceph/operator.go]: > `ceph.rook.io/filesystem`

agree

File: pkg/operator/ceph/provisioner/provisioner.go (hunk 2 above)
Comment (caisan) [reply pkg/operator/ceph/provisioner/provisioner.go]: > What about having a different code file for cephfs provisioner in the same package?

i think it's unnecessary, because cephfs provisioner or rbd(block) provisioner are belongs to storage provisioner, it's reasonable in the same provisioner.go file.

hope your replay @rohan47

File: Documentation/ceph-filesystem.md [hunk 4]
```diff
@@ -139,6 +139,44 @@ After creating it with `kubectl create -f kube-registry.yaml`, you now have a do
...
 
+## Consume the Shared File System: K8s PVC
+
+In order to comsume shared file system by using PVC, you need create storageclass firstly. Here is the sample of storageclass:
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: rook-ceph-fs
+provisioner: ceph.rook.io/filesystem
```
Comment (travisn) [inline Documentation/ceph-filesystem.md]: The storage class will need something to identify which ceph cluster since it's possible to run multiple ceph clusters with the same operator. For example, see the [clusterNamespace](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/storageclass.yaml#L26) property for the rbd storageclass

File: images/ceph/__init__.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2017 The Kubernetes Authors.
```
Comment (travisn) [inline images/ceph/__init__.py]: why is this file needed?

File: images/ceph/cephfs_provisioner.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
+#!/usr/bin/env python
```
Comment (travisn) [inline images/ceph/cephfs_provisioner.py]: Where is this script triggered? I don't see it called anywhere.

Comment (caisan) [reply images/ceph/cephfs_provisioner.py]: In ```images/ceph/Dockerfile```,  ```images/ceph/cephfs_provisioner.py``` is copy to ```/usr/local/bin/cephfs_provisioner```.
triggered at ```cmd := exec.Command(provisionCmd, args...)``` in provisioner.go

File: Documentation/ceph-filesystem.md (hunk 4 above)
Comment (caisan) [reply Documentation/ceph-filesystem.md]: @travisn add ```clusterNamespace``` would be great. thanks.

**Comment by caisan** [comment]
> In Rook v1.1 focus is moving towards the ceph-csi driver, which would have dynamic provisioning for cephfs. I'm not sure we want to invest in this feature if it's immediately going to be replaced by the csi driver already in the same release. Have you tried the csi driver to confirm if it will work for you?
> Thanks for the significant contribution, I'm sure you've spent a bunch of time getting this working.

no, i have not tried the csi driver.  so, could you please tell me the detail about how to work with csi driver?  thanks a lot. Or you mean, will the ceph-csi driver be implemented in the next release(v1.1) ?

**Comment by galexrt** [comment]
@caisan There is a documentation page for using the Ceph CSI driver: https://rook.io/docs/rook/v1.0/ceph-csi-drivers.html

The goal is to use the Ceph CSI as default and also automatically set it up through the Rook Ceph operator in v1.1.

**Comment by caisan** [comment]
> @caisan There is a documentation page for using the Ceph CSI driver: https://rook.io/docs/rook/v1.0/ceph-csi-drivers.html
> 
> The goal is to use the Ceph CSI as default and also automatically set it up through the Rook Ceph operator in v1.1.

cool, i skip through the doc, there is something different from the way of creating rbd pvc now in the rook. so, if it means the code of running rbd volume provisioner by goroutine in operator.go will be delete when release v1.1 ?
this PR(#3264 ) amis to run a cephfs provisioner as same as the rbd provisioner in operator. so i think it is reasonable to do this PR.

**Comment by travisn** [comment]
@caisan The Rook documentation for dynamic provisioning of filesystem has been updated and can be seen here in the pre-release 1.1 documentation: https://rook.io/docs/rook/v1.1/ceph-filesystem.html

Could you take a look at that and see if it would meet your requirements?

**Comment by galexrt** [comment]
@caisan Are you available to update this PR?

**Comment by galexrt** [comment]
@caisan Ping are you available to update this PR or should we close it?

**Comment by travisn** [comment]
Or with the dynamic provisioning for cephfs in the csi driver shall we close this?

**Comment by leseb** [comment]
Closing this due to inactivity. Feel free to re-open. Thanks.
